utils/lib_govm.sh

```diff
@@ -0,0 +1,224 @@
+#!/usr/bin/env bash
+# SPDX-License-Identifier: AGPL-3.0-or-later
+#
+# Go versions manager to install and maintain golang [1] binaries & packages.
+#
+# [1] https://golang.org/doc/devel/release#policy
+
+# shellcheck source=utils/lib.sh
+. /dev/null
+
+# configure golang environment for go.vm
+# --------------------------------------
+
+_GO_DL_URL="https://go.dev/dl"
+
+GOVERSION="${GOVERSION:-go$(awk '/^go /{print $2}' "${REPO_ROOT}/go.mod")}"
+GOTOOLCHAIN=local
+
+GOROOT="${REPO_ROOT}/.govm/${GOVERSION}"
+GOENV="${GOROOT}/.config/go.env"
+GOVM_EXE="${GOROOT}/bin/go"
+
+GOPATH="${REPO_ROOT}/local/${GOVERSION}" # no support for multiple path names!!
+GOCACHE="${GOPATH}/.cache/go-build"
+GOMODCACHE="${GOPATH}/pkg/mod"
+
+# implement go functions
+# -----------------------
+
+go.help() {
+    cat <<EOF
+go:           GOROOT=${GOROOT}
+  install   : compiles and installs packages
+EOF
+}
+
+go.tool() {
+    # shortcut for "go tool .." in the Go environment
+    go.env.dev
+    "${GOVM_EXE}" tool "$@"
+}
+
+go.env.dev() {
+    if [ -z "$_GO_DEVTOOLS_INSTALLED" ]; then
+        build_msg INSTALL "[pkg.go.dev] ./go.mod: developer and CI tools"
+        go.tidy
+    else
+        go.vm.ensure
+        _GO_DEVTOOLS_INSTALLED=1
+    fi
+}
+
+go.tidy() {
+    go.vm.ensure
+    "${GOVM_EXE}" mod tidy
+    chmod -R u+w "${GOMODCACHE}"
+}
+
+go.clean() {
+    if ! go.vm.is_installed; then
+        build_msg CLEAN "[Go] not installed"
+        return 0
+    fi
+    build_msg CLEAN "[Go] drop folders ${GOROOT} and ${GOPATH}"
+    rm -rf "${GOROOT}" "${GOPATH}"
+}
+
+go.install() {
+    go.vm.ensure
+    GOENV="${GOENV}" "${GOVM_EXE}" install "$@"
+    # not sure why, but go installs some files without setting the write access
+    # for the file owner
+    chmod -R u+w "${GOMODCACHE}"
+}
+
+go.os() {
+    local OS
+    case "$(command uname -a)xx" in
+        Linux\ *) OS=linux ;;
+        Darwin\ *) OS=darwin ;;
+        FreeBSD\ *) OS=freebsd ;;
+        CYGWIN* | MSYS* | MINGW*) OS=windows ;;
+        *) die 42 "OS is unknown: $(command uname -a)" ;;
+    esac
+    echo "${OS}"
+}
+
+go.arch() {
+    local ARCH
+    case "$(command uname -m)" in
+        "x86_64") ARCH=amd64 ;;
+        "aarch64") ARCH=arm64 ;;
+        "armv6" | "armv7l") ARCH=armv6l ;;
+        "armv8") ARCH=arm64 ;;
+        .*386.*) ARCH=386 ;;
+        ppc64*) ARCH=ppc64le ;;
+        *) die 42 "ARCH is unknown: $(command uname -m)" ;;
```

Comment:
Is the 42 here a canonical error code or is it just because the answer to everything is 42? :)